File: 337-house-robber-iii/house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<pair<TreeNode*,int>,int> m;
    int fun(TreeNode* node,int taken)
    {
        if(node==nullptr)
        {
            return 0;
        }
        if(m.find({node, taken})!=m.end())
        {
            return m[{node,taken}];
        }
        int take=0,skip=0;
        if(taken==0)
        {
            take = fun(node->left,1) + node->val + fun(node->right,1);
            skip = fun(node->left,0) + fun(node->right,0);
        }
        else{
             skip = fun(node->left,0) + fun(node->right,0); 
        }
        return m[{node,taken}] = max(take,skip);
    }
    int rob(TreeNode* root) {
        return fun(root,0);
    }
};

File: 337-house-robber-iii/house-robber-iii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    Map<TreeNode, Integer> mpCanSelect = new HashMap<>();
    Map<TreeNode, Integer> mpMustSkip = new HashMap<>();

    private int helper(TreeNode root, int i) {
        if (root == null) {
            return 0;
        }

        if (i == 0 && mpCanSelect.containsKey(root)) {
            return mpCanSelect.get(root);
        }
        if (i == 1 && mpMustSkip.containsKey(root)) {
            return mpMustSkip.get(root);
        }

        int select = 0, skip = 0;
        if (i == 0) {
            select = root.val + helper(root.right, 1) + helper(root.left, 1);
        }
        skip = helper(root.right, 0) + helper(root.left, 0);
        int ans = Math.max(skip, select);
        if (i == 0) {
            mpCanSelect.put(root, ans);
        } else {
            mpMustSkip.put(root, ans);
        }

        return ans;
    }

    public int rob(TreeNode root) {
        mpCanSelect.clear();
        mpMustSkip.clear();
        return helper(root, 0);
    }
}